scripts/integrity_check.py

```diff
@@ -0,0 +1,141 @@
+#!/usr/bin/env python3
+#
+# This file is part of mbed TLS (https://tls.mbed.org)
+#
+# Copyright (c) 2018, Arm Limited, All Rights Reserved
+#
+# Purpose
+#
+# This script requires python 3. This script checks for undesired changes
+# to the source code, including file permission changes, presence of tabs,
+# non-Unix line endings, trailing whitespace, and TODO comments.
```

Comment:
I have new a feature request: check for non-ASCII characters. There are currently no non-ASCII characters in our code base.

If we decide to allow non-ASCII characters, then please at the very least ensure that files don't start with `\xef\xbb\xbf` (the UTF-8 “BOM”). Yes, I'm asking for it because it happened and it isn't visually apparent.

Reply:
Darryl let me know that I'd flubbed my check, we do have comments with people's names containing non-ASCII characters. So I'm revising my request to just forbidding the UTF-8 BOM.

Reply:
@gilles-peskine-arm @dgreen-arm If I run the script on a file saved as `UTF-16` in `gedit`, the script chokes and fails with 
```
hanno@wuehlmaus:~/mbedtlsX$ ./scripts/integrity_check.py 
Traceback (most recent call last):
  File "./scripts/integrity_check.py", line 217, in <module>
    run_main()
  File "./scripts/integrity_check.py", line 212, in run_main
    integrity_check.check_files()
  File "./scripts/integrity_check.py", line 190, in check_files
    issue_to_check.check_file_for_issue(filepath)
  File "./scripts/integrity_check.py", line 84, in check_file_for_issue
    if not f.read().endswith("\n"):
  File "/usr/lib/python3.5/codecs.py", line 321, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte
```

Reply:
You need to open the file in binary. Python 3 supposedly makes this easier than Python 2.

Reply:
I can change the script to always open the file in binary to avoid these issues. Would we want to check for the UTF-16 and UTF-32 BOM's as well?

Reply:
If a file was in UTF-16 or UTF-32, then a BOM would be desirable. I suppose we could check that no file that's supposed to be a text file contains a null byte, or more generally control characters not among `\n\r\t\f`. That would in practice detect any file saved as UTF-16 or UTF-32. I don't think we've ever had this particular problem, but we might as well check this while we're at it.